create mongodb indexes article (#236)
prisma/dataguide: content/08-mongodb/09-mongodb-indexes.mdx

</PrismaOutlinks>

## When should you use an index
Keeping with the our encyclopedia analogy, one might thing to have an index line for every word in the book. If it is always faster to use an index, then this seems beneficial. However, as you can imagine, the more lines of words in an index then the larger the book becomes. At some point the size of the book to accomodate indexing every word becomes ineffective. There are many words such as "the" or "because" that are less useful to search for than "hippopotamus".

This is similar to an index in MongoDB and databases in general. While yes, it is faster to have an index for any possible query, there simply are queries that don't need indexing. Like the size of a book, adding too many indexes to a database also takes up space and has a detrimental effect on write operations on the database if not kept in check.

Indexes are an incredibly useful way to optimize querying your database and are recommended. Knowing when to use them is important, so making sure that you are adding them on database fields that are often being queried will keep your reads efficient without negatively impacting database size and write efficiency.

> 

Reviewer: It might be useful to add a quick sentence explaining that you want to index fields that are explicitly used as selection criteria when retrieving records.  For instance, if you have an ID column or you often search for records by email address, you'd want to add indexes to those fields.

The basic syntax for seeing all of your collection's indexes is:

```
db.COLLECTION_NAME.getIndexes()
``` 

Using the previous example from when we created an index, the following displays the method and what it would return.

```
db.mycoll.getIndexes()
```
with a return of:

```
[
   {
      "v" : 2,
      "key" : {
         "country" : 1
      },
      "name" : "country"
   }
 ]
```
Index information includes the keys and options used to create the index.

<PrismaOutlinks>

If you're interested in using full text indexes with MongoDB, Prisma has the [`fullTextIndex` preview feature](https://www.prisma.io/docs/concepts/components/prisma-schema/indexes#full-text-indexes-mysql-and-mongodb) which allows you to easily migrate full text indexes into a Prisma schema for type safety and protection against validation errors.
</PrismaOutlinks>  

## How to understand index performance
Now with the ability to create and check what indexes exist on your collection, you will want to see if your indexes are performing the way you expect.

> To begin the example, we will work with the `sample_mflix` database and the `comments` collection which has ~50.3k documents. This is a sample collecton provided by MongoDB simulating a datastore of movie and tv comments. 

Reviewer: Can we link to this database?

Author: I can link to the MongoDB university where I got it from. It's their free education for MongoDB

```
db.comments.find( { "name" : "Ramsay Bolton" } )
```
Now, if we append [MongoDB's explain plan](https://docs.mongodb.com/manual/reference/explain-results/) to the query, we will see the performance of this query.

```
db.comments.find( { "name" : "Ramsay Bolton" } ).explain("executionStats")
```
This results in the following:

```
{
  queryPlanner: {
    plannerVersion: 1,
    namespace: 'sample_mflix.comments',
    indexFilterSet: false,
    parsedQuery: { name: { '$eq': 'Ramsay Bolton' } },
    winningPlan: {
      stage: 'COLLSCAN',
      filter: { name: { '$eq': 'Ramsay Bolton' } },
      direction: 'forward'
    },
    rejectedPlans: []
  },
  executionStats: {
    executionSuccess: true,
    nReturned: 273,
    executionTimeMillis: 23,
    totalKeysExamined: 0,
    totalDocsExamined: 50303,
    executionStages: {
      stage: 'COLLSCAN',
      filter: { name: { '$eq': 'Ramsay Bolton' } },
      nReturned: 273,
      executionTimeMillisEstimate: 6,
      works: 50305,
      advanced: 273,
      needTime: 50031,
      needYield: 0,
      saveState: 50,
      restoreState: 50,
      isEOF: 1,
      direction: 'forward',
      docsExamined: 50303
    }
  }
}
```

> There are several key results to focus on in this output. First, we can see in the `winningPlan` that the `stage` for this query is a `COLLSCAN`. This means that a collection scan occurred to commplete this query with a `totalDocsExamined` of 50,303 and `executionTimeMillis` of 23 milliseconds. The query had to examine every document in the collection and took 23ms even though `nReturned` was only 273 documents. While 23ms may not sound like much, this can get much longer for a collection housing one million documents.

Reviewer: ```suggestion
There are several key results to focus on in this output. First, we can see in the `winningPlan` that the `stage` for this query is a `COLLSCAN`. This means that a collection scan occurred to complete this query with a `totalDocsExamined` of 50,303 and `executionTimeMillis` of 23 milliseconds. The query had to examine every document in the collection and took 23ms even though `nReturned` was only 273 documents. While 23ms may not sound like much, this can get much longer for a collection housing one million documents.
```

> If querying on `name` is going to be a recurrent query for the application accessing this collection, we may want to create an index on this field. To do so we write the following:

Reviewer: ```suggestion
If querying on `name` is going to be a recurrent pattern for the application accessing this collection, we may want to create an index on this field. To do so we write the following:
```

```
db.comments.createIndex( {"name":1} )
```

> If we take the same query with the explain planfrom before:

Reviewer: ```suggestion
If we take the same query with the explain plan from before, we get:
```

```
db.comments.find( { "name" : "Ramsay Bolton" } ).explain("executionStats")
```

> we get:

Reviewer: ```suggestion
```

      }
    },
    rejectedPlans: []
  },
  executionStats: {
    executionSuccess: true,
    nReturned: 273,
    executionTimeMillis: 0,
    totalKeysExamined: 273,
    totalDocsExamined: 273,
    executionStages: {
      stage: 'FETCH',
      nReturned: 273,
      executionTimeMillisEstimate: 0,
      works: 274,
      advanced: 273,
      needTime: 0,
      needYield: 0,
      saveState: 0,
      restoreState: 0,
      isEOF: 1,
      docsExamined: 273,
      alreadyHasObj: 0,
      inputStage: {
        stage: 'IXSCAN',
        nReturned: 273,
        executionTimeMillisEstimate: 0,
        works: 274,
        advanced: 273,
        needTime: 0,
        needYield: 0,
        saveState: 0,
        restoreState: 0,
        isEOF: 1,
        keyPattern: { name: 1 },
        indexName: 'name_1',
        isMultiKey: false,
        multiKeyPaths: { name: [] },
        isUnique: false,
        isSparse: false,
        isPartial: false,
        indexVersion: 2,
        direction: 'forward',
        indexBounds: { name: [ '["Ramsay Bolton", "Ramsay Bolton"]' ] },
        keysExamined: 273,
        seeks: 1,
        dupsTested: 0,
        dupsDropped: 0
      }
    }
  }
}
```

> Comparing the same fields as the un-indexed query we see that the `stage` from the `winningPlan` is now `IXSCAN` which indicates an index was used.

Reviewer: I read the `stage` in `winningPlan` as being `FETCH`.  Are we supposed to look at the `stage` in `winningPlan.inputStage` instead?

Additionally, we see that `totalDocsExamined` is now just the 273 documents where `name` is `"Ramsay Bolton"` and not the entire 50,303 documents. This increased efficiency can be see especially now with `executionTimeMillis` totaling 0ms. 

> Analyzing the explain plan on your most important queries will show you the performance of your index or if one needs to be created to increase the efficiency of your application. 

Reviewer: ```suggestion
Analyzing the explain plan on your most important queries will show you the performance of your index or highlight when one needs to be created to increase the efficiency of your application. 
```

## How to drop an index

> While the explain plan may show the need for an index, it can also do the opposite. For example, if an index is no longer needed or not adding that much improved performance, it may be in your best interest to drop that index and reserve the space or preserve some write performance. 

Reviewer: ```suggestion
While the explain plan may show the need for an index, it can also do the opposite. For example, if an index is no longer needed or not adding that much improved performance, it may be in your best interest to drop that index to reserve the space and gain some write performance. 
```

To drop an index on your collection, the basic syntax is as follows using the `dropIndexes()` method:

```
db.COLLECTION_NAME.dropIndex( { "FIELD_NAME": 1 } )
```
If we wanted to drop our `country` index example from earlier, we would write the following:

```
db.mycoll.dropIndex( { "country":1 } )
```

## Conclusion

> Efficiently querying your database leads to an improved user experience of your application. Additionally those working with the data for analysis or other internal efforts will have faster performance and ease of working with the database. Knowing how to index and how indexes work is key to achieving query efficiency. 

Reviewer: ```suggestion
Efficiently querying your database leads to an improved user experience of your application. Additionally those using with the data for analytics or other internal efforts will have faster performance and ease of working with the database. Knowing how to index and how indexes work is key to achieving query efficiency. 
```

I changed "working" -> "using" here just to mix it up from the "working" later in the sentence.

> We covered the basics of creating, analyzing, and dropping indexes in MongoDB. Knowing these index foundations will provide the right base for continuing on to more advanced indexing with MongoDB.

Reviewer: Maybe a lead in like "In this guide..." or something would work here.
